[DNF-5] Tests for the Package class

lukash opened the pull request:
Plus minor improvements.

lukash committed:
tests: Create RepoFixture to share the code for repository setup

lukash committed:
tests: Add a class enabling std collection assertions

Plus a helper function for converting ReldepList to a vector of strings,
which enables easy assertions on ReldepLists.

lukash committed:
tests: Add tests for libdnf::rpm::Package

lukash committed:
rpm/checksum: Fix construction with a nullptr

lukash committed:
rpm/package: Call the correct metod for get_requires()

j-mracek commented on test/libdnf/rpm/test_package.cpp:
```diff
@@ -0,0 +1,346 @@
+libdnf::rpm::Package RpmPackageTest::get_pkg(const std::string &nevra) {
+    libdnf::rpm::SolvQuery query(sack.get());
+    query.ifilter_nevra(libdnf::sack::QueryCmp::EQ, {nevra});
+    CPPUNIT_ASSERT_MESSAGE("get_pkg(\"" + nevra + "\"): no package found.", query.size() > 0);
```
What about to use assert `query.size() == 1`?

lukash replied:
Yeah.. that will probably be better...

j-mracek commented on test/libdnf/rpm/test_package.cpp:
```diff
@@ -0,0 +1,346 @@
+void RpmPackageTest::test_equality() {
+    auto package1 = get_pkg("CQRlib-1.1.1-4.fc29.x86_64");
+    auto package2 = get_pkg("CQRlib-1.1.1-4.fc29.x86_64");
+    CPPUNIT_ASSERT(package1 == package2);
```
Is it possible tu use `CPPUNIT_ASSERT_EQUAL`?

lukash replied:
Not here, `CPPUNIT_ASSERT_EQUAL` doesn't use `operator==`, it's got its own set of classes for this (you can see the commit for asserting std collections for a rough idea, I didn't dig much further). Also, I am testing the `operator==` itself, so that's what I should be calling.

Thinking about it now, I should also test `operator==` for when the packages aren't equal and `operator!=` when they are equal...

j-mracek commented on test/libdnf/rpm/test_package.cpp:
```diff
@@ -0,0 +1,346 @@
+
+
+void RpmPackageTest::test_get_id() {
+    get_pkg("CQRlib-1.1.1-4.fc29.x86_64").get_id();  // Can't really test for the value of the ID
```
In future we can load particular rpm in certain order - like packages from commandline. Then we will know ids.
For the present time we can make a test that `get_pkg("CQRlib-1.1.1-4.fc29.x86_64").get_id() != 0`.

j-mracek commented on test/libdnf/rpm/test_package.cpp:
```diff
@@ -0,0 +1,346 @@
+        std::string("CQRlib-1.1.1-4.fc29.src.rpm"),
+        get_pkg("CQRlib-1.1.1-4.fc29.x86_64").get_sourcerpm()
+    );
+    // TODO is it correct it's empty?
```
I suggest that it is correct that it is empty. SRC rpm should have no src entry/. Also the data are originated from libsolv and metadata created by createrepo_c, therefore we have to keep it like it is. Please could you remove TODO note?

j-mracek commented on test/libdnf/rpm/test_package.cpp:
```diff
@@ -0,0 +1,346 @@
+
+
+void RpmPackageTest::test_get_build_host() {
+    // TODO is it correct it's empty?
```
To check if it is correct please open metadata and try to find all related information to the package. Libsolv cannot generate data that are not present in xml.

lukash replied:
Looking into the repo's `primary.xml`, there's `<rpm:buildhost>rohel</rpm:buildhost>`. So that would be non-empty. Looking at definition of `SOLVABLE_BUILDHOST`, which is the libsolv "type" of the build host attribute, it's `"solvable:buildhost"`. I'm not sure what is going on between `primary.xml` and libsolv's loaded repo data though, that would take some time to dig through libsolv. Should I do it or are we happy with the test the way it is for now?

j-mracek replied:
I suggest that we should found some package that will return a different value from an empty string. Or do you think that the get_build_host() works incorrectly?

lukash replied:
I've tried taking a package from the distro (`libdnf`) and call the `get_build_host()` on it, it's empty as well. I don't know whether this is incorrect behavior.

j-mracek replied:
I know that it will take a time, but please could you look in libsolv how data are parsed, or if we use the functionality correctly? The method that looks at `SOLVABLE_BUILDHOST` is new in dnf-5, therefore we can also drop it if it is not working correctly.

j-mracek replied:
@lukash Please could you report the status?

lukash replied:
@j-mracek I'm not working on this, as I've said in https://github.com/rpm-software-management/libdnf/pull/1006#issuecomment-672761009 and also on one of our meetings. As I've said, I'd prefer to merge this now and fix the remaining TODOs in a separate PR when I get to it.

j-mracek replied:
From comments I am not 100% sure what you found in libsolv. Is functionality to get SOLVABLE_BUILDHOST in libsolv implemented and working? At least I can see that installed RPMs have data about build_host (`rpm -qi dnf`).

lukash replied:
I didn't look into libsolv at all. Given how complex the internals are I expect it will take some time to find out what's going on. The only information I found is mentioned in https://github.com/rpm-software-management/libdnf/pull/1006#discussion_r466454553

j-mracek replied:
Thanks, I just discussed it with @dmach and according to his investigation the functionality is not implemented in libsolv. Only flag was created and reserved.

lukash replied:
Well, I guess that explains it :slightly_smiling_face:

j-mracek replied:
It means we cannot merge it like it is.

j-mracek replied:
I suggest to create an issue on libsolv, comment out `get_build_host()` from code, add ToDo reference on the issue, and remove `RpmPackageTest::test_get_build_host()`.

lukash replied:
We can do that. Though the method apparently has been broken in dnf 4 as well and nobody ever minded that. The build host information seems to be practically useless (and anyone needing it can also query the rpm directly). So we could drop it from the API and focus on core functionality.

If we want to pursue this though, I'd keep the method in place, even the test. Create the libsolv issue and when it starts working, the test should actually start failing and thus notify us :slightly_smiling_face:

j-mracek replied:
get_build_host() is not in dnf-4 code. I suggest to comment the function out and create an issue. Keeping not functional code in API is dangerous.

j-mracek replied:
The function was describe in API-DN5 document created by @dmach. He had a good argument. Why we ship information in metadata/header that we are unable to provide to end/API users.

lukash replied:
Libsolv issue: https://github.com/openSUSE/libsolv/issues/400

j-mracek commented on test/libdnf/rpm/test_package.cpp:
```diff
@@ -0,0 +1,346 @@
+
+void RpmPackageTest::test_get_files() {
+    const auto files = get_pkg("CQRlib-1.1.1-4.fc29.x86_64").get_files();
+    // TODO no files in test packages
```
You can try following packages or modify other packages:
```
CQRlib-1.1.1-4.fc29.src ['/CQRlib-1.1.1-4.fc29.spec']
CQRlib-devel-1.1.2-16.fc29.src ['/CQRlib-devel-1.1.2-16.fc29.spec']
abcde-2.9.2-1.fc29.src ['/abcde-2.9.2-1.fc29.spec']
basesystem-11-6.fc29.src ['/basesystem-11-6.fc29.spec']
dwm-6.1-1.src ['/dwm-6.1-1.fc29.spec']
fedora-release-29-1.src ['/fedora-release-29-1.spec']
filesystem-3.9-2.fc29.src ['/filesystem-3.9-2.fc29.spec']
flac-1.3.2-8.fc29.src ['/flac-1.3.2-8.fc29.spec']
glibc-2.28-9.fc29.src ['/glibc-2.28-9.fc29.spec']
glibc-2.28-9.fc29.x86_64 ['/etc/ld.so.conf', '/var/db']
http-parser-2.4.0-1.fc29.src ['/http-parser-2.4.0-1.fc29.spec']
kernel-4.18.16-300.fc29.src ['/kernel-4.18.16-300.fc29.spec']
lame-3.100-4.fc29.src ['/lame-3.100-4.fc29.spec']
lz4-1.7.5-2.fc26.src ['/lz4-1.7.5-2.fc26.multilib.spec']
ninja-build-1.8.2-5.fc29.src ['/ninja-build-1.8.2-5.fc29.spec']
nodejs-1:5.12.1-1.fc29.src ['/nodejs-5.12.1-1.fc29.spec']
postgresql-9.6.5-1.fc29.src ['/postgresql-9.6.5-1.fc29.spec']
setup-2.12.1-1.fc29.src ['/setup-2.12.1-1.fc29.spec']
wget-1.19.5-5.fc29.src ['/wget-1.19.5-5.fc29.spec']
yum-3.4.3-0.src ['/yum-3.4.3-0.fc29.spec']
```

j-mracek commented on test/libdnf/rpm/test_package.cpp:
```diff
@@ -0,0 +1,346 @@
+
+void RpmPackageTest::test_get_requires_pre() {
+    const auto prereqs = get_pkg("filesystem-3.9-2.fc29.x86_64").get_requires_pre();
+    // TODO no prereqs in test packages
```
What about to modify one RPM to have something to test?

j-mracek commented on test/libdnf/rpm/test_package.cpp:
```diff
@@ -0,0 +1,346 @@
+void RpmPackageTest::test_get_prereq_ignoreinst() {
+    const auto ipr = get_pkg("npm-1:5.12.1-1.fc29.x86_64").get_prereq_ignoreinst();
+    // TODO no prereq_ignoreinst (called Ipr: in libsolv testcases) in test packages
+    const std::vector<std::string> expected = {};
```
What about to modify one RPM to have something to test?

lukash replied:
Do you know how to put this in a spec file? I'm unable to find this in documentation anywhere.

j-mracek replied:
Probably this explanation could help:
```
/**
 * dnf_package_get_prereq_ignoreinst
 * @pkg: a #DnfPackage instance.
 *
 * Returns safe to remove Requires(pre) and Requires(post) dependencies
 * of the package. Safe to remove means that the dependency is not needed by
 * Requires(preun) nor by Requires(postun).
 * The package must be installed otherwise empty list is returned.
 *
 * Returns: A #DnfReldepList
 *
 * Since: 0.45.1
 */
```

lukash replied:
Yeah, it does. The package needs to be installed though.

j-mracek commented on test/libdnf/rpm/test_package.cpp:
```diff
@@ -0,0 +1,346 @@
+
+void RpmPackageTest::test_get_regular_requires() {
+    const auto reg_reqs = get_pkg("filesystem-3.9-2.fc29.x86_64").get_regular_requires();
+    // TODO doesn't test a difference between requires() and this, since there are no prereqs
```
Please could you add required data for the test?

j-mracek commented on test/libdnf/rpm/test_package.cpp:
```diff
@@ -0,0 +1,346 @@
+void RpmPackageTest::test_get_enhances() {
+    const auto enhances = get_pkg("abcde-2.9.2-1.fc29.noarch").get_enhances();
+    // TODO no enhances in test packages
+    const std::vector<std::string> expected = {};
```
Please could you add required data for the test?

j-mracek commented on test/libdnf/rpm/test_package.cpp:
```diff
@@ -0,0 +1,346 @@
+
+
+void RpmPackageTest::test_get_baseurl() {
+    CPPUNIT_ASSERT_EQUAL(std::string(), get_pkg("CQRlib-1.1.1-4.fc29.x86_64").get_baseurl());
```
Please could you investigate how to not have baseurl empty? Then set an additional test that will compare it. The comparison with empty string is not a good approach. Also some features could be tested in CI_DNF_STACK where more options are available.

j-mracek replied:
Please what's the status?

lukash replied:
The test is correct, `Package::get_baseurl()` returns a non-empty string only if the `xml:base` attribute is present in the package's location tag in `primary.xml`: `<location xml:base="http://example.com" href="x86_64/foo-1.0-1.x86_64.rpm"/>`. This isn't the case for the testing data we have now and adding this attribute is done using `createrepo_c`: `createrepo_c --baseurl "http://example.com" ...`.

This would best be done after the testing data are refactored, right now it's the generated data that are stored in git.

j-mracek replied:
But you already added or updated test data for tests therefore it can finished completely.

lukash replied:
What do you mean? That I can add this data as well? Have a look at https://github.com/rpm-software-management/libdnf/pull/1006/commits/8cc7aa4d21e9e149f58e82564c998d16caf11919. I've included the line I used to regenerate the repo data in the commit message. The value for the baseurl here needs to be part of that line.

Also, doing it this way, either all the packages have the `xml:base` set or they don't meaning I can't easily test both it being there and not.

I can add a simple script to git with the `createrepo_c` line so that it's at least part of git (come think of it I should do it anyway), but it's still just putting patches over patches on the already insufficient solution.

j-mracek replied:
Or you can add another repository for the test, just small one. As I already mentioned in this PR, the support of multiple repositories in test is useful.

j-mracek commented on test/libdnf/rpm/test_package.cpp:
```diff
@@ -0,0 +1,346 @@
+
+void RpmPackageTest::test_get_hdr_checksum() {
+    // TODO is it correct it's empty?
+    CPPUNIT_ASSERT_EQUAL(std::string(""), get_pkg("CQRlib-1.1.1-4.fc29.x86_64").get_hdr_checksum().get_checksum());
```
HEADER checksum is available only for installed packages, therefore it is correct.

j-mracek replied:
Please the note must be in format TODO, then add tests for installed packages

lukash replied:
Again, not sure what you mean. I've added "TODO" to the comment (the snippet github now shows is outdated too).

j-mracek commented on test/libdnf/rpm/test_package.cpp:
```diff
@@ -0,0 +1,346 @@
+
+void RpmPackageTest::test_get_hdr_end() {
+    // TODO what is this API method for? do we really want it?
+    CPPUNIT_ASSERT_EQUAL(6208llu, get_pkg("CQRlib-1.1.1-4.fc29.x86_64").get_hdr_end());
```
Please could you investigate it in dnf4/libdnf?

lukash replied:
It is on the API, but unused. It's hard to imagine any API user needing that...

j-mracek replied:
It was part of python API, therefore we cannot say it is unused.

lukash replied:
We cannot say that, but we can make an educated guess the information obtained from the method is almost certainly useless to dnf API users and drop the method. On the rare occasion it would be needed, we can always add it on request.

j-mracek replied:
I discussed the functionality with dmach and we agree that there are several cases when the function/information could be useful. Please could you remove TODO?

lukash replied:
Sure. Out of curiosity, what are the usecases?

j-mracek replied:
You can download only rpm header, or only payload when you know where header ends.

lukash committed:
rpm/solv_sack: Add dump_debugdata()

Dumps the libsolv debug data (although there's no transaction, it does
dump the loaded repository data, which is useful for debugging).

lukash committed:
test/libdnf/rpm: Add source rpms to the testing repo

The rpms were in the repodata, but not present in the repo. The repodata
are regenerated for consistency using the following command:

createrepo_c --no-database --simple-md-filenames --revision=1550000000 .

lukash committed:
test/libdnf/rpm/test_solv_query: Fix some of the CPPUNIT_ASSERT macros

Change the macros to CPPUNIT_ASSERT_EQUAL for more clarity when
debugging.

lukash committed:
test/libdnf/rpm: Add packages for the Package class unittests

Repodata generated with:
createrepo_c --no-database --simple-md-filenames --revision=1550000000 .

lukash commented:
@j-mracek I've pushed fixes for most of your comments. I'm not sure when I'll have time to do the investigation for the remaining ones. FWIW it might be better to merge this now and do the remaining work later (there are still plenty other TODOs).

I'd also like to point out the binary RPMs as well as the repodata are in a **very** messy state right now and something needs to be done about that. I only added the ones in this PR on your request and IMO it's just making matters worse in the long run.

j-mracek commented on include/libdnf/rpm/solv_sack.hpp:
```diff
@@ -127,6 +127,8 @@ class SolvSack {
     /// Creates system repository and loads it into SolvSack. Only one system repository can be in SolvSack.
     void create_system_repo(bool build_cache = false);
 
+    void dump_debugdata(const std::string & dir);
```
I am not sure whether we want to keep the functionality exactly in this place, because there will be an overlap with goal debug. Please could you add TODO that we have to reconsider it in future?

lukash replied:
I can add the TODO; I've noticed the dump is normally used with a transaction to dump its data as well, but I found it quite useful just to dump the repositories, when you may not have any transaction at all.

j-mracek commented on test/libdnf/rpm/repo_fixture.cpp:
```diff
@@ -0,0 +1,65 @@
+#include <filesystem>
+
+
+void RepoFixture::setUp() {
```
How do you want to support a multiple repository environment?

lukash replied:
Not sure what you mean exactly, but I didn't plan for anything like that. I've just created the class just to share (and deduplicate) the repository setup code. If there are test suites that require different repo setup, the code can be changed to accommodate for that when needed.

j-mracek replied:
I can have two repository for tests. Then I need to set which one I want to load. See dnf, hawkey unittests.

lukash replied:
When you have the two repos, you're free to adjust the code to make it work :slightly_smiling_face: These are just unittests, not a public API... I've just fixed the multiple duplication of the code and set up the fixture for it.

j-mracek commented on test/libdnf/rpm/repo_fixture.cpp:
```diff
@@ -0,0 +1,65 @@
+    );
+}
+
+void RepoFixture::dump_debugdata() {
```
Why you have it, but you do not use it?

lukash replied:
It's there ready to use when someone needs to debug a test.

lukash committed:
rpm/package: Comment out Package::get_build_host()

The attribute is not supported by libsolv:
https://github.com/openSUSE/libsolv/issues/400

lukash committed:
test/libdnf/rpm: Add a repo with a package with baseurl

Used for the Package class tests.

lukash committed:
test/libdnf/rpm: Add a script to generate repodata

lukash commented:
@j-mracek I've commented out the `get_build_host()` and added a TODO for that, and added a second repo with the baseurl for the package set.

j-mracek commented:
@lukash Perfekt
